### PR TITLE
metered mpsc channel

### PR DESCRIPTION
**Pull request by drahnr**
Adds the ability to track the amount of messages sitting in the channel to be processed. To be extended to have added `Total`s of messages (potentially per message type) also tracked as needed.

### CODE REVIEW

File: node/subsystem-util/src/tracked.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+					let msg = Msg::default();
+					assert_eq!(tx.queue_count(), 0);
+					tx.try_send(msg).unwrap();
+					assert_eq!(tx.queue_count(), 1);
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+				},
+				async move {
+					assert_eq!(rx.queue_count(), 4);
```
Comment (coriolinus) [inline node/subsystem-util/src/tracked.rs:205]: I don't think that this test can work properly unless you have some kind of external synchronization which ensures that we don't attempt this assertion until the other future has run to completion.

**Review by coriolinus — COMMENTED** [review]
Looks like CI is failing due to code external to this PR. I do like this code, but I don't understand how the test mentioned below can pass.

File: node/subsystem-util/src/tracked.rs (hunk 1 above)
Comment (drahnr) [reply node/subsystem-util/src/tracked.rs]: Thanks, mere luck this did not blow up.

File: node/subsystem-util/src/metered.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+					let msg = Msg::default();
+					assert_eq!(tx.meter().queue_count(), 0);
+					tx.try_send(msg).unwrap();
+					assert_eq!(tx.meter().queue_count(), 1);
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+					tx.try_send(msg).unwrap();
+				},
+				async move {
+					Delay::new(Duration::from_millis(100)).await;
```
Comment (coriolinus) [inline node/subsystem-util/src/metered.rs:221]: Delaying here will work almost all of the time, but it's not 100% reliable. If you create a mutex outside the scope of the `block_on`, move the lock into the other async block, then `lock().await` here, you have a 100% reliable synchronization mechanism, which is likely to complete faster than 100ms.

Comment (drahnr) [reply node/subsystem-util/src/metered.rs:221]: Yeah, but good enough - it's just a test :) both futures start as ready and a couple of atomic ops will never take 100ms.

Comment (montekki) [reply node/subsystem-util/src/metered.rs:338]: you can probably sync on a `oneshot`?

File: node/subsystem-util/src/metered.rs [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+	/// The sender component, tracking the number of items
+	/// sent across it.
+	#[derive(Debug)]
+	pub struct UnboundedMeteredSender<T> {
+		meter: Meter,
+		inner: mpsc::UnboundedSender<T>,
+	}
+
+	impl<T> Clone for UnboundedMeteredSender<T> {
```
Comment (montekki) [inline node/subsystem-util/src/metered.rs:228]: it looks like it is equal to `#[derive(Clone)]`?

File: node/subsystem-util/src/metered.rs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+/// The sender component, tracking the number of items
+/// sent across it.
+#[derive(Debug)]
+pub struct MeteredSender<T> {
+	meter: Meter,
+	inner: mpsc::Sender<T>,
+}
+
+impl<T> Clone for MeteredSender<T> {
```
Comment (montekki) [inline node/subsystem-util/src/metered.rs:103]: also a `derive`?

File: node/overseer/src/lib.rs [hunk 5]
```diff
@@ -1242,15 +1244,15 @@ where
 		CG: Subsystem<OverseerSubsystemContext<CollationGenerationMessage>> + Send,
 		CP: Subsystem<OverseerSubsystemContext<CollatorProtocolMessage>> + Send,
 	{
-		let (events_tx, events_rx) = mpsc::channel(CHANNEL_CAPACITY);
+		let (events_tx, events_rx) = metered::channel(CHANNEL_CAPACITY, "overseer_events");
```
Comment (montekki) [inline node/overseer/src/lib.rs:1247]: How exactly are these supposed to be used? It would probably be useful to automatically create a metric in the constructor for each channel created

**Review by montekki — APPROVED** [review]
lgtm; just some probable improvements

File: node/subsystem-util/src/metered.rs (hunk 3 above)
Comment (drahnr) [reply node/subsystem-util/src/metered.rs:228]: Let me try again, initially that did not work.

File: node/subsystem-util/src/metered.rs (hunk 4 above)
Comment (drahnr) [reply node/subsystem-util/src/metered.rs:103]: C & P

File: node/overseer/src/lib.rs (hunk 5 above)
Comment (drahnr) [reply node/overseer/src/lib.rs:1247]: That was one option but I decided to allow having a `Meter` accessor that has access to the `Arc<Atomic*>` so snapshots can be made at arbitrary points in time.

Comment (montekki) [reply node/overseer/src/lib.rs:1278]: in any case probably useful logging/metrics of some sort would be really good within this PR, like the periodic metric updates in overseer loop or smething

Comment (drahnr) [reply node/overseer/src/lib.rs:1278]: What do you have in mind? Just stats every 10secs or a histogram with ascii bars?

Comment (montekki) [reply node/overseer/src/lib.rs:1278]: just -per-1-sec prometheus metrics with a `Gauge`?

Comment (drahnr) [reply node/overseer/src/lib.rs:1278]: Oh, I see

Comment (drahnr) [reply node/overseer/src/lib.rs:1278]: I'll skip that for now, the histogram sampling should be enough and there would make the instantiation of a channel failable or the metric could not be registered and would be silently dropped. Not sure either is good.

File: node/overseer/src/lib.rs [hunk 6]
```diff
@@ -1242,23 +1275,39 @@ where
...
+			let meter_to_overseer = to_overseer_rx.meter().clone();
+			let metronome_metrics = metrics.clone();
+			let metronome = Metronome::new(std::time::Duration::from_millis(137))
+			.for_each(move |_| {
+				metronome_metrics.channel_fill_level_snapshot(meter_from_overseer.queue_count(), meter_to_overseer.queue_count());
+
+				async move {
+					()
+				}
+			});
```
Comment (bkchr) [inline node/overseer/src/lib.rs:1299]:
```suggestion
			.for_each(move |_| async move {
				metronome_metrics.channel_fill_level_snapshot(meter_from_overseer.queue_count(), meter_to_overseer.queue_count());
			});
```

Comment (drahnr) [reply node/overseer/src/lib.rs:1299]: This won't work, since the closure has to retain the ownership, it's a `Fn` not an `FnOnce`.

Comment (bkchr) [reply node/overseer/src/lib.rs:1299]: Hmm okay :P

File: node/subsystem-util/src/metered/unbounded.rs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metered variant of bounded mpsc channels to be able to extract metrics.
```
Comment (ordian) [inline node/subsystem-util/src/metered/unbounded.rs:17]:
```suggestion
//! Metered variant of unbounded mpsc channels to be able to extract metrics.
```

File: node/subsystem-util/src/metered/mod.rs [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+impl Meter {
+	/// Count the number of items queued up inside the channel.
+	pub fn queue_count(&self) -> usize {
+		// when obtaining we don't care much about off by one
+		// accuracy
+		self.fill.load(Ordering::Relaxed)
+	}
+
+	/// Obtain the name of the channel `Sender` and `Receiver` pair.
+	pub fn name(&self) -> &'static str {
```
Comment (ordian) [inline node/subsystem-util/src/metered/mod.rs:51]: how names are going to be used? given the `snake_case` convention looks like somehow in prometheus, but that will like create a [cardinality issue](https://www.robustperception.io/cardinality-is-key)

Comment (drahnr) [reply node/subsystem-util/src/metered/mod.rs:51]: Currently they are not used as such, the idea was to provide a unique identifier that _could_ be used for proemtheus i.e. - currently it is mere convenience and the metrics for prometheus must be implemented deliberately.

File: node/overseer/src/lib.rs [hunk 9]
```diff
@@ -1121,6 +1130,30 @@ impl metrics::Metrics for Metrics {
 				)?,
 				registry,
 			)?,
+			channel_fill_level_from_overseer: prometheus::register(
+				prometheus::Histogram::with_opts(
+					prometheus::HistogramOpts {
+						common_opts: prometheus::Opts::new(
+							"channel_fill_level_from_overseer",
```
Comment (ordian) [inline node/overseer/src/lib.rs:1137]: we use `parachain_` prefix for all metrics, looks like a few slipped already

File: node/subsystem-util/src/lib.rs [hunk 10]
```diff
@@ -50,6 +50,7 @@ use streamunordered::{StreamUnordered, StreamYield};
 use thiserror::Error;
 
 pub mod validator_discovery;
+pub mod metered;
```
Comment (ordian) [inline node/subsystem-util/src/lib.rs:53]: since we're splitting subsystem-util in #2214, maybe we could put it in a separate crate to avoid another batch of merge conflicts?

**Review by ordian — APPROVED** [review]

File: node/subsystem-util/src/lib.rs (hunk 10 above)
Comment (drahnr) [reply node/subsystem-util/src/lib.rs:53]: Sure thing!

**Comment by drahnr** [comment]
Feel free to bikeshed on the crate name `metered-channel`. Otherwise this is ready to merge.

File: node/overseer/src/lib.rs [hunk 11]
```diff
@@ -1134,7 +1134,7 @@ impl metrics::Metrics for Metrics {
 				prometheus::Histogram::with_opts(
 					prometheus::HistogramOpts {
 						common_opts: prometheus::Opts::new(
-							"parachain_channel_fill_level_from_overseer",
+							"overseer_channel_fill_level_from_overseer",
```
Comment (ordian) [inline node/overseer/src/lib.rs:1137]: is there a reason to use a different prefix for overseer? we already mention its name at the end
also prometheus recomments to use a suffix describing the unit:
https://prometheus.io/docs/practices/naming/

Comment (drahnr) [reply node/overseer/src/lib.rs:1137]: @ordian If you add a change to the name, feel free to suggest one. I added overseer since there already was one and it seemed more fitting than `parachain`.
